File: apps/mobile/android/app/src/main/java/com/aurelia/app/ui/LoginViewModel.kt
package com.aurelia.app.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aurelia.app.storage.SessionStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.aurelia_core.AppException
import uniffi.aurelia_core.AuthRequest
import uniffi.aurelia_core.BackendProvider
import uniffi.aurelia_core.authenticate
import uniffi.aurelia_core.detectProvider

class LoginViewModel(
  private val sessionStore: SessionStore,
) : ViewModel() {
  private val mutableState =
    MutableStateFlow(
      LoginState(useDynamicColor = sessionStore.getUseDynamicColor()),
    )
  val state: StateFlow<LoginState> = mutableState

  fun updateServerUrl(value: String) {
    mutableState.update { it.copy(serverUrl = value, detectedProvider = null) }
  }

  fun updateUsername(value: String) {
    mutableState.update { it.copy(username = value) }
  }

  fun updatePassword(value: String) {
    mutableState.update { it.copy(password = value) }
  }

  fun updateProviderSelection(selection: LoginProviderSelection) {
    mutableState.update { it.copy(providerSelection = selection) }
  }

  fun detectProviderNow() {
    val current = mutableState.value
    if (current.serverUrl.isBlank()) {
      mutableState.update { it.copy(error = "Server URL is required to detect provider") }
      return
    }

    mutableState.update { it.copy(isDetectingProvider = true, error = null) }

    viewModelScope.launch(Dispatchers.IO) {
      try {
        val detectedProvider = detectProvider(current.serverUrl.trim())
        mutableState.update {
          it.copy(
            isDetectingProvider = false,
            detectedProvider = detectedProvider,
          )
        }
      } catch (error: AppException) {
        mutableState.update {
          it.copy(isDetectingProvider = false, error = error.message ?: "Provider detection failed")
        }
      } catch (error: Exception) {
        mutableState.update {
          it.copy(isDetectingProvider = false, error = "Unexpected error: ${error.message}")
        }
      }
    }
  }

  fun toggleDynamicColor(enabled: Boolean) {
    sessionStore.setUseDynamicColor(enabled)
    mutableState.update { it.copy(useDynamicColor = enabled) }
  }

  fun submit() {
    val current = mutableState.value
    if (current.serverUrl.isBlank() || current.username.isBlank() || current.password.isBlank()) {
      mutableState.update { it.copy(error = "All fields are required") }
      return
    }

    mutableState.update { it.copy(isSubmitting = true, error = null) }

    viewModelScope.launch(Dispatchers.IO) {
      try {
        val resolvedProvider = when (current.providerSelection) {
          LoginProviderSelection.JELLYFIN -> BackendProvider.JELLYFIN
          LoginProviderSelection.NAVIDROME -> BackendProvider.NAVIDROME
          LoginProviderSelection.AUTO -> current.detectedProvider ?: detectProvider(current.serverUrl.trim())
        }

        val response = authenticate(
          AuthRequest(
            provider = resolvedProvider,
            serverUrl = current.serverUrl.trim(),
            username = current.username,
            password = current.password,
            deviceId = sessionStore.getDeviceId(),
          ),
        )
        sessionStore.save(
          serverUrl = current.serverUrl.trim(),
          userId = response.userId,
          token = response.token,
          username = current.username,
          provider = resolvedProvider,
        )
        mutableState.update {
          it.copy(
            isSubmitting = false,
            token = response.token,
            userId = response.userId,
            detectedProvider = resolvedProvider,
          )
        }
      } catch (error: AppException) {
        mutableState.update {
          it.copy(isSubmitting = false, error = error.message ?: "Login failed")
        }
      } catch (error: Exception) {
        mutableState.update {
          it.copy(isSubmitting = false, error = "Unexpected error: ${error.message}")
        }
      }
    }
  }
}

File: apps/mobile/ios/Aurelia/ViewModels/LoginViewModel.swift
import AureliaCore
import Foundation
import Observation
import UIKit

enum LoginProviderSelection: String, CaseIterable, Identifiable {
    case auto
    case jellyfin

    var id: String { rawValue }

    var title: String {
        switch self {
        case .auto:
            return "Auto"
        case .jellyfin:
            return "Jellyfin"
        }
    }
}

@Observable
final class LoginViewModel: @unchecked Sendable {
    var serverUrl = ""
    var username = ""
    var password = ""
    var providerSelection: LoginProviderSelection = .auto
    var detectedProvider: BackendProvider?
    var isDetectingProvider = false
    var isSubmitting = false
    var error: String?

    // Set after successful login
    var token: String?
    var userId: String?

    private let sessionStore = SessionStore.shared

    func detectProviderNow() {
        guard let normalizedServerUrl = ServerURLNormalizer.normalizeForServer(raw: serverUrl),
              ServerURLNormalizer.isValidServerURL(normalizedServerUrl)
        else {
            error = "Enter a valid server URL"
            return
        }

        isDetectingProvider = true
        error = nil
        serverUrl = normalizedServerUrl

        Task.detached { [serverUrl = normalizedServerUrl] in
            do {
                let provider = try await detectProvider(serverUrl: serverUrl)
                await MainActor.run { [self] in
                    isDetectingProvider = false
                    detectedProvider = provider
                }
            } catch {
                await MainActor.run { [self] in
                    isDetectingProvider = false
                    self.error = error.localizedDescription
                }
            }
        }
    }

    func submit() {
        guard !serverUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !username.trimmingCharacters(in: .whitespaces).isEmpty,
              !password.isEmpty
        else {
            error = "All fields are required"
            return
        }

        guard let normalizedServerUrl = ServerURLNormalizer.normalizeForServer(raw: serverUrl),
              ServerURLNormalizer.isValidServerURL(normalizedServerUrl)
        else {
            error = "Enter a valid server URL"
            return
        }

        isSubmitting = true
        error = nil
        serverUrl = normalizedServerUrl
        let providerSelection = self.providerSelection
        let detectedProvider = self.detectedProvider

        let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? "aurelia-ios-\(UUID().uuidString)"

        Task.detached {
            [serverUrl = normalizedServerUrl, username = self.username, password = self.password, deviceId, providerSelection, detectedProvider] in
            do {
                let resolvedProvider: BackendProvider = .jellyfin

                let response = try await authenticate(
                    request: AuthRequest(
                        provider: resolvedProvider,
                        serverUrl: serverUrl,
                        username: username,
                        password: password,
                        deviceId: deviceId
                    )
                )
                await MainActor.run {
                    let sessionStore = SessionStore.shared
                    sessionStore.save(
                        serverUrl: serverUrl,
                        userId: response.userId,
                        token: response.token,
                        username: username,
                        provider: resolvedProvider
                    )
                }

                await MainActor.run { [self] in
                    isSubmitting = false
                    token = response.token
                    userId = response.userId
                    self.detectedProvider = resolvedProvider
                }
            } catch {
                await MainActor.run { [self] in
                    isSubmitting = false
                    self.error = error.localizedDescription
                }
            }
        }
    }
}
